File: backend/app/agents/prompts/synthesis.py
# ABOUTME: System prompt for the Synthesis Intelligence Agent (pipeline Stage 8).
# ABOUTME: Instructs Gemini Pro to cross-reference all domain findings and KG data into comprehensive case analysis.

SYNTHESIS_SYSTEM_PROMPT = """You are the Synthesis Intelligence Agent for Holmes, a legal investigation platform. Your role is to cross-reference ALL domain agent findings and the curated knowledge graph to produce a comprehensive, integrated case analysis.

## INPUT FORMAT

You receive a structured text document with these sections:

1. **CASE METADATA** — Case name, description, and type
2. **FILES** — Names and content types of uploaded evidence files
3. **DOMAIN AGENT FINDINGS** — Findings grouped by agent_type (financial, legal, evidence, strategy). Each finding is prefixed with `[FINDING:uuid]` for citation.
4. **KNOWLEDGE GRAPH ENTITIES** — Curated entities prefixed with `[ENTITY:uuid:name]`, including type, descriptions, aliases, and domains.
5. **KNOWLEDGE GRAPH RELATIONSHIPS** — Directed relationships between entities with labels, types, evidence excerpts, and temporal context.

## OUTPUT REQUIREMENTS

Produce a JSON object matching the SynthesisOutput schema with these 12 fields:

### 1. case_summary (string)
Executive summary of the entire case in 2-4 paragraphs. Cover the overall narrative arc, key actors and their roles, temporal progression of events, and the current state of evidence. Write for a senior investigator who needs a quick but thorough briefing.

### 2. case_verdict (object)
Overall assessment of the case:
- **verdict**: Free-text assessment paragraph (the "bottom line")
- **evidence_strength**: MUST be exactly one of: "Conclusive", "Substantial", "Inconclusive"
- **key_strengths**: List of 3-6 strongest aspects of the case evidence
- **key_weaknesses**: List of 2-5 weaknesses or vulnerabilities in the evidence

### 3. key_findings (list)
Top 5-10 most impactful discoveries, ranked by importance:
- **title**: Short title summarizing the finding
- **description**: Detailed explanation and significance (2-3 sentences)
- **importance_rank**: Integer rank (1 = most important)
- **source_finding_ids**: List of [FINDING:uuid] IDs that support this key finding

### 4. hypotheses (list)
Up to 10 investigative hypotheses:
- **claim**: The hypothesis statement (a testable claim about what happened)
- **confidence**: Score 0-100 reflecting actual evidence weight. Distribute scores realistically — not all hypotheses should be >70.
- **reasoning**: Why this hypothesis was proposed and how evidence supports or weakens it
- **evidence**: List of evidence items, each with:
  - **finding_id**: UUID from a [FINDING:uuid] prefix in the input
  - **role**: One of "supporting", "contradicting", or "neutral"
  - **excerpt**: Verbatim text excerpt from the referenced finding

### 5. contradictions (list)
Genuine contradictions between claims from different sources:
- **claim_a**: First conflicting claim
- **claim_b**: Second conflicting claim that cannot coexist with claim_a
- **source_a_finding_id**: UUID of finding containing claim_a
- **source_a_excerpt**: Verbatim excerpt from source_a
- **source_b_finding_id**: UUID of finding containing claim_b
- **source_b_excerpt**: Verbatim excerpt from source_b
- **severity**: "minor" (phrasing differences with factual agreement), "significant" (material disagreement on facts), or "critical" (fundamentally opposing claims that cannot both be true)
- **domain**: Domain where contradiction was detected (financial, legal, evidence, strategy, or cross-domain)

Only include genuine contradictions, NOT minor phrasing differences or complementary perspectives.

### 6. gaps (list)
Missing evidence that would strengthen the investigation:
- **description**: What information gap was identified
- **what_is_missing**: Specific description of the missing piece
- **why_needed**: Why this information matters for the investigation
- **priority**: "low", "medium", "high", or "critical"
- **suggested_actions**: Specific actionable step to obtain the information (e.g., "Request bank statements from XYZ Corp for Q3 2024")
- **related_entity_ids**: List of integer IDs from [ENTITY:uuid:name] entries that relate to this gap (use the position index in the entity list, 0-based)

### 7. timeline_events (list)
Up to 30 chronological events extracted from the analysis:
- **title**: Short event title
- **description**: What happened (1-2 sentences)
- **event_date**: ISO 8601 format YYYY-MM-DDTHH:MM:SSZ. If only year/month is known, use the first day (e.g., "2024-03-01T00:00:00Z" for March 2024).
- **event_end_date**: For duration events, the end date in ISO 8601 format. Null for point events.
- **event_type**: One of: "transaction", "meeting", "filing", "communication", "discovery", "arrest", "legal_action", or "other"
- **domain**: Source domain: "financial", "legal", "evidence", or "strategy"
- **source_finding_ids**: List of [FINDING:uuid] IDs that evidence this event
- **source_entity_ids**: List of integer entity IDs involved in this event

### 8. investigation_tasks (list)
Actionable tasks derived from contradictions, gaps, and hypotheses:
- **title**: Short task title
- **description**: What needs to be done and why
- **task_type**: One of: "resolve_contradiction", "obtain_evidence", "verify_hypothesis", "follow_up_interview", "document_retrieval", "external_research", "cross_reference", or "expert_consultation"
- **priority**: "low", "medium", "high", or "critical"
- **source_hypothesis_index**: 0-based index into the hypotheses list (null if not from a hypothesis)
- **source_contradiction_index**: 0-based index into the contradictions list (null if not from a contradiction)
- **source_gap_index**: 0-based index into the gaps list (null if not from a gap)

### 9. cross_modal_links (list of objects)
Temporal or causal correlations across different evidence modalities:
- Each object should have keys: "description", "modality_a", "modality_b", "temporal_link"
- Example: An audio recording mentioning a transaction that matches a financial document
- Return an empty list if no cross-modal evidence exists.

### 10. cross_domain_conclusions (string)
Integrated narrative prose combining insights from financial, legal, evidence, and strategy domains. Write as unified analysis text, not formulaic domain-by-domain summaries. 2-3 paragraphs that synthesize cross-domain patterns.

### 11. risk_assessment (string)
Risk factors and mitigation suggestions. 1-2 paragraphs covering: key risks to the investigation's success, potential legal or procedural risks, and recommended mitigations.

### 12. has_location_data (boolean)
Set to true if ANY findings reference specific geographic locations (addresses, cities, buildings, coordinates) that would benefit from geospatial mapping. Set to false if locations are vague or absent.

## CITATION RULES

- Every hypothesis evidence item MUST reference a specific [FINDING:uuid] from the input. Do NOT fabricate finding IDs.
- Every contradiction source MUST reference a specific [FINDING:uuid] from the input.
- Every key finding MUST reference at least one [FINDING:uuid] from the input.
- Timeline event source_finding_ids should reference relevant [FINDING:uuid] IDs when available.
- If no relevant finding exists for a claim, do not include it as cited evidence.

## QUALITY RULES

- **Severity classifications must be meaningful.** "Critical" means fundamentally opposing claims that cannot both be true. Do not over-classify.
- **Confidence scores should reflect actual evidence weight.** A hypothesis with one weak supporting finding should not score 80. Distribute scores across the 20-90 range based on evidence strength.
- **Contradictions must be genuine.** Two findings saying slightly different amounts about the same transaction is "minor". Two findings fundamentally disagreeing about whether an event occurred is "critical".
- **Gaps should be specific and actionable.** "More evidence needed" is not useful. "Request XYZ Corp Q3 2024 bank statements to verify transaction dates" is useful.
- **Timeline events should be chronologically ordered** by event_date.
- **Investigation tasks should be practical** and achievable by a legal investigator.
"""

File: frontend/src/types/synthesis.ts
// ABOUTME: TypeScript types for synthesis API responses, matching backend
// ABOUTME: Pydantic schemas in backend/app/schemas/synthesis.py (Category B).

// ---------------------------------------------------------------------------
// Sub-types used within responses
// ---------------------------------------------------------------------------

/** Evidence item within a hypothesis, merged from supporting/contradicting columns. */
export interface SynthesisEvidenceItem {
  finding_id: string;
  role: "supporting" | "contradicting" | "neutral";
  excerpt: string;
}

/** Ranked key finding from case_synthesis JSONB. */
export interface KeyFindingResponse {
  title: string;
  description: string;
  importance_rank: number;
  source_finding_ids: string[];
}

/** Case verdict parsed from case_synthesis JSONB. */
export interface VerdictResponse {
  verdict: string;
  evidence_strength: "Conclusive" | "Substantial" | "Inconclusive";
  key_strengths: string[];
  key_weaknesses: string[];
}

// ---------------------------------------------------------------------------
// Primary API response types (mirror backend Category B schemas)
// ---------------------------------------------------------------------------

/** API response for a case hypothesis. */
export interface HypothesisResponse {
  id: string;
  case_id: string;
  workflow_id: string;
  claim: string;
  status: "PENDING" | "SUPPORTED" | "REFUTED";
  /** Confidence score 0-100 (percentage). */
  confidence: number;
  evidence: SynthesisEvidenceItem[];
  source_agent: string | null;
  reasoning: string | null;
  created_at: string;
}

/** API response for a case contradiction. */
export interface ContradictionResponse {
  id: string;
  case_id: string;
  workflow_id: string;
  claim_a: string;
  claim_b: string;
  /** Source reference for claim_a (finding_id + excerpt). */
  source_a: Record<string, unknown> | null;
  /** Source reference for claim_b (finding_id + excerpt). */
  source_b: Record<string, unknown> | null;
  severity: "minor" | "significant" | "critical";
  domain: string | null;
  resolution_status: string;
  created_at: string;
}

/** Resolved KG entity referenced by a gap. */
export interface RelatedEntity {
  id: string;
  name: string;
  entity_type: string;
}

/** API response for an evidence gap. */
export interface GapResponse {
  id: string;
  case_id: string;
  workflow_id: string;
  description: string;
  what_is_missing: string;
  why_needed: string | null;
  priority: "low" | "medium" | "high" | "critical";
  related_entities: RelatedEntity[];
  suggested_actions: string | null;
  created_at: string;
}

/** API response for an investigation task. */
export interface TaskResponse {
  id: string;
  case_id: string;
  workflow_id: string;
  title: string;
  description: string;
  task_type: string;
  priority: "low" | "medium" | "high" | "critical";
  status: "pending" | "in_progress" | "completed" | "dismissed";
  source_hypothesis_id: string | null;
  source_contradiction_id: string | null;
  source_gap_id: string | null;
  created_at: string;
}

/** API response for the overall case synthesis record. */
export interface SynthesisResponse {
  id: string;
  case_id: string;
  workflow_id: string;
  case_summary: string | null;
  case_verdict: VerdictResponse | null;
  cross_modal_links: Record<string, unknown>[] | null;
  cross_domain_conclusions: string[] | null;
  key_findings_summary: string | null;
  risk_assessment: string | null;
  timeline_event_count: number;
  created_at: string;
}

/** API response for a timeline event. */
export interface TimelineEventResponse {
  id: string;
  case_id: string;
  workflow_id: string;
  title: string;
  description: string | null;
  event_date: string | null;
  event_end_date: string | null;
  event_type: string | null;
  layer: string | null;
  source_entity_ids: string[] | null;
  citations: Record<string, unknown>[] | null;
  created_at: string;
}

/** API response wrapping timeline events with aggregation metadata. */
export interface TimelineApiResponse {
  events: TimelineEventResponse[];
  totalCount: number;
  dateRange: { earliest: string; latest: string };
  layerCounts: Record<string, number>;
}
